File: app/src/main/java/com/donald/musictheoryapp/Question/CheckBoxQuestion.java
package com.donald.musictheoryapp.Question;

import com.donald.musictheoryapp.Utils.JSONArrayUtil;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class CheckBoxQuestion extends Question
{
    public static class Answer implements Question.Answer
    {
        public Boolean userAnswer;
        public boolean correctAnswer;

        public Answer(Boolean userAnswer, boolean correctAnswer)
        {
            this.userAnswer = userAnswer;
            this.correctAnswer = correctAnswer;
        }

        public JSONObject toJSON() throws JSONException
        {
            JSONObject object = new JSONObject();
            object.put("user_answer", userAnswer == null ? JSONObject.NULL : userAnswer);
            object.put("correct_answer", correctAnswer);
            return object;
        }

        public static Answer fromJSON(JSONObject object) throws JSONException
        {
            return new Answer(
                object.isNull("user_answer") ? null : object.getBoolean("user_answer"),
                object.getBoolean("correct_answer")
            );
        }

        @Override
        public boolean correct()
        {
            return userAnswer != null && userAnswer == correctAnswer;
        }
    }

    public static final String CHECK_ANSWER = "check";
    public static final String CROSS_ANSWER = "cross";

    public Answer[] answers;

    public CheckBoxQuestion() {}

    @Override
    public int points()
    {
        int points = 0;
        for(Answer answer : answers)
        {
            if(answer.correct()) points++;
        }
        return points;
    }

    @Override
    public void acceptVisitor(QuestionVisitor visitor) { visitor.visit(this); }

    public static CheckBoxQuestion fromJSON(JSONObject object, QuestionGroup group) throws JSONException
    {
        CheckBoxQuestion question = new CheckBoxQuestion();
        question.number = object.getInt("number");
        question.group = group;
        question.descriptions = JSONArrayUtil.descriptions(object);
        JSONArray answerJSONArray = object.getJSONArray("answers");
        question.answers = new Answer[answerJSONArray.length()];
        for(int i = 0; i < question.answers.length; i++)
        {
            question.answers[i] = Answer.fromJSON(answerJSONArray.getJSONObject(i));
        }
        return question;
    }
}

File: app/src/main/java/com/donald/musictheoryapp/question/CheckBoxQuestion.kt
package com.donald.musictheoryapp.question

import kotlin.Throws
import com.donald.musictheoryapp.util.getDescriptions
import com.donald.musictheoryapp.util.getInputHint
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

class CheckBoxQuestion(
    number: Int,
    descriptions: List<Description>,
    inputHint: String?,
    val answers: List<Answer>
) : ChildQuestion(number, descriptions, inputHint) {

    override val points: Int
        get() = answers.count { it.correct }
    override val maxPoints = answers.size
    override val correctAnswerCount = 1

    override fun acceptVisitor(visitor: QuestionVisitor) {
        visitor.visit(this)
    }

    override fun copyNew(): CheckBoxQuestion {
        return CheckBoxQuestion(
            number,
            List(descriptions.size) { i -> descriptions[i].copy() },
            inputHint,
            List(answers.size) { i -> answers[i].copyNew() }
        )
    }

    override fun toPartialJson(): JSONObject {
        val jsonObject = JSONObject()
        jsonObject.put(QUESTION_TYPE, Type.CHECK_BOX.ordinal)
        val array = JSONArray()
        for (answer in answers) {
            array.put(answer.toJson())
        }
        jsonObject.put("answers", array)
        return jsonObject
    }

    companion object {

        const val CHECK_ANSWER = "check"
        const val CROSS_ANSWER = "cross"
        @Throws(JSONException::class)
        fun fromJson(jsonObject: JSONObject): CheckBoxQuestion {
            val question = CheckBoxQuestion(
                number = jsonObject.getInt("number"),
                descriptions = jsonObject.getDescriptions(),
                inputHint = jsonObject.getInputHint(),
                answers = jsonObject.getJSONArray("answers").run {
                    List(this.length()) { index ->
                        Answer.fromJson(this.getJSONObject(index))
                    }
                }
            )
            return question
        }

    }

    class Answer(var userAnswer: Boolean?, val correctAnswer: Boolean) : ChildQuestion.Answer {

        override val correct: Boolean
            get() = userAnswer == correctAnswer

        override fun copyNew(): Answer {
            return Answer(null, correctAnswer)
        }

        @Throws(JSONException::class)
        fun toJson(): JSONObject {
            val jsonObject = JSONObject()
            jsonObject.put("user_answer", if (userAnswer == null) JSONObject.NULL else userAnswer)
            jsonObject.put("correct_answer", correctAnswer)
            return jsonObject
        }

        companion object {

            @Throws(JSONException::class)
            fun fromJson(jsonObject: JSONObject): Answer {
                return Answer(
                    if (jsonObject.isNull("user_answer")) null else jsonObject.getBoolean("user_answer"),
                    jsonObject.getBoolean("correct_answer")
                )
            }

        }

    }
}